tools/gen_dir_structure.py

```diff
@@ -0,0 +1,311 @@
+"""
+
+from argparse import (
+    ArgumentParser,
+    Namespace
+)
+from datetime import (
+    datetime,
+    timedelta
+)
+from pathlib import Path
+from typing import List
+
+import os
+import sys
+
+from stem.descriptor.networkstatus import KeyCertificate, NetworkStatusDocumentV3
+
+from gen_fresh_dirinfo import (
+    consensus_validate_signatures,
+    fetch_certificates,
+    fetch_latest_consensus,
+    generate_certificate,
+    generate_churninfo,
+    generate_customized_consensus,
+    InvalidConsensus,
+    LOGGER
+)
+
+DIRNAME_AUTHORITY_PRIVATE = "private"
+DIRNAME_AUTHORITY_PUBLIC = "public"
+
+FILENAME_AUTHORITY_KEY_ID = "authority_identity_key"
+FILENAME_AUTHORITY_KEY_SIGNATURE = "authority_signing_key"
+FILENAME_AUTHORITY_FINGERPRINT = "authority.txt"
+FILENAME_CERTIFICATE = "certificate.txt"
+FILENAME_CHURN = "churn.txt"
+FILENAME_CONSENSUS = "consensus.txt"
+FILENAME_MICRODESCRIPTORS = "microdescriptors.txt"
+
+# Feel free to modify these values as you see fit.
+CERTIFICATE_VALIDITY_MONTHS = 12
+CONSENSUS_NUMBER_ROUTERS = 120
+CONSENSUS_VALIDITY_DAYS = 7
+
+# Feel free to modify these values as you see fit.
+# Note: These parameters are necessary to generate a directory authority, but their values are not
+# important.
+AUTHORITY_NAME = "C4DT"
+AUTHORITY_HOSTNAME = "c4dt.org"
+AUTHORITY_IP_ADDRESS = "128.178.32.16"
+AUTHORITY_DIR_PORT = 80
+AUTHORITY_OR_PORT = 443
+AUTHORITY_CONTACT = "https://www.c4dt.org"
+
+
+class InconsistentDirectoryStructure(Exception):
+    """
+    The directory structure is not consistant.
+    """
+
+
+def fetch_valid_consensus() -> NetworkStatusDocumentV3:
+    """
+    Retrieve a fresh consensus from the network and verify its signature.
+
+    :raises InvalidConsensus: the consensus is invalid
+    :return: the fresh consensus
+    """
+    consensus = fetch_latest_consensus()
+    v3idents = [auth.v3ident for auth in consensus.directory_authorities]
+    key_certificates = fetch_certificates(v3idents)
+    if not consensus_validate_signatures(consensus, key_certificates):
+        raise InvalidConsensus("Validation of the consensus' signature failed.")
+
+    return consensus
+
+
+def ensure_dir_rw_access(directory: Path) -> None:
+    """
+    Ensures that a directory exists and that we have read/write access to it.
+
+    :param directory: directory we want to ensure to have these properties
+    :raises InconsistentDirectoryStructure: The directory does not have these properties.
+    """
+    if not directory.exists():
+        directory.mkdir()
+
+    if not directory.is_dir():
+        raise InconsistentDirectoryStructure(f"File {directory} should be a directory.")
+
+    if not os.access(directory, os.R_OK | os.W_OK):
+        raise InconsistentDirectoryStructure(f"Directory {directory} do not have enough permission.")
+
+
+def ensure_file_r_access(filepath: Path) -> None:
+    """
+    Ensures that a file exists and that we have read/write access to it.
```

Comment:
the method name and implementation suggests that we only need _read_ access

Reply:
We only need read access when issuing a custom consensus, but we would need write access if we need to build an authority or to renew the certificate.

When building a new authority, its directory is created, so it should not pose problems, but when reissuing the certificate, it might indeed cause problem. I am going to add a check.